compiler/daml-lf-verify/src/DA/Daml/LF/Verify/Context.hs

```diff
@@ -0,0 +1,812 @@
+lookupVar x = getEnv >>= \case
+  EnvVG{..} -> return $ elem (SkolVar x) _envvgskol
+  EnvCG{..} -> return $ elem (SkolVar x) _envcgskol
+  EnvS{..} -> return $ elem (SkolVar x) _envsskol
+
+-- | Lookup a record project in the environment. Returns a boolean denoting
+-- whether or not the record projection has been skolemised.
+lookupRec :: MonadEnv m ph
+  => ExprVarName
+  -- ^ The expression variable on which is being projected.
+  -> FieldName
+  -- ^ The field name which is being projected.
+  -> m Bool
+lookupRec x f = do
+  skols <- getEnv >>= \case
+    EnvVG{..} -> return _envvgskol
+    EnvCG{..} -> return _envcgskol
+    EnvS{..} -> return _envsskol
+  let fields = [ fs | SkolRec y fs <- skols, x == y ]
+  if not (null fields)
+    then return (elem f $ head fields)
+    else return False
+
+-- | Lookup a value name in the environment. Returns its (partially) evaluated
+-- definition, together with the updates it performs.
+lookupVal :: MonadEnv m ph
+  => Qualified ExprValName
+  -- ^ The value name to lookup.
+  -> m (Expr, UpdateSet ph)
+lookupVal val = do
+  vals <- getEnv >>= \case
+    EnvVG{..} -> return _envvgvals
+    EnvCG{..} -> return _envcgvals
+    EnvS{..} -> return _envsvals
+  case lookupValInHMap vals val of
+    Just res -> return res
+    Nothing -> throwError (UnknownValue val)
+
+-- | Lookup a choice name in the environment. Returns a function which, once
+-- self, this and args have been instantiated, returns the set of updates it
+-- performs.
+lookupChoice :: MonadEnv m ph
+  => Qualified TypeConName
+  -- ^ The template name in which this choice is defined.
+  -> ChoiceName
+  -- ^ The choice name to lookup.
+  -> m (Expr -> Expr -> Expr -> UpdateSet ph)
+lookupChoice tem ch = do
+  chs <- getEnv >>= \case
+    EnvVG{..} -> return HM.empty
+    EnvCG{..} -> return _envcgchs
+    EnvS{..} -> return _envschs
+  case lookupChoInHMap chs (qualObject tem) ch of
+    Nothing -> throwError (UnknownChoice ch)
+    Just (_,_,_,upd) -> return upd
+
+-- | Lookup a data type definition in the environment.
+lookupDataCon :: MonadEnv m ph
+  => TypeConName
+  -- ^ The data constructor to lookup.
+  -> m DefDataType
+lookupDataCon tc = do
+  dats <- getEnv >>= \case
+    EnvVG{..} -> return _envvgdats
+    EnvCG{..} -> return _envcgdats
+    EnvS{..} -> return _envsdats
+  case HM.lookup tc dats of
+    Nothing -> throwError (UnknownDataCons tc)
+    Just def -> return def
+
+-- | Lookup a contract id in the environment. Returns the variable its fetched
+-- contract is bound to.
+lookupCid :: MonadEnv m ph
+  => Expr
+  -- ^ The contract id to lookup.
+  -> m ExprVarName
+lookupCid exp = do
+  cid <- expr2cid exp
+  cids <- getEnv >>= \case
+    EnvVG{..} -> return _envvgcids
+    EnvCG{..} -> return _envcgcids
+    EnvS{..} -> return _envscids
+  case HM.lookup cid cids of
+    Nothing -> throwError $ UnknownCid cid
+    Just var -> return var
+
+-- TODO: There seems to be something wrong with the PackageRef in Qualified.
```

Comment:
Are you aware that there is `PRSelf` for references into the same package? You might want to rewrite all these references into `PRImport <own-package-id>` during package loading.